### PR TITLE
enh(auth): Add Authentication Conditions to OpenID Connect Authentication

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

This PR add the possibility to use Authentication Conditions on OpenID.

While the authentication condition mode is enable, we request the provider to get authentications conditions.
The response is parsed using the configured attribute path in Centreon openid configuration.
If the parsing matching and an authorized values is found, so the user is connected. Else he is redirected to a new page `/authentication-denied`

**Fixes** # MON-15005

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.04.x
- [ ] 21.10.x
- [ ] 22.04.x
- [x] 22.10.x (master)

<h2> How this pull request can be tested ? </h2>

- Configure an OpenID Provider in Centreon and enable authentication conditions
- Configure your attribute path
- Configure your authorized values
- Configure the endpoint to call to get the provider authentication conditions
- Try to authenticate yourself using OpenID

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: src/Core/Security/ProviderConfiguration/Domain/OpenId/Exceptions/OpenIdConfigurationException.php [hunk 1]
```diff
@@ -100,4 +100,19 @@ public static function missingMandatoryParameters(array $missingParameters): sel
...
+     * Exception thrown when the Authentication Endpoints is not valid.
+     *
+     * @param string $endpoint
+     * @return self
+     */
+    public static function invalidAuthenticationConditionsEndpoint(string $endpoint): self
+    {
+        return new self(_(sprintf(
+            'The authentication conditions endpoint should match your introspection or user information endpoints:'
+                . '%s given',
```
Comment (kduret) [inline src/Core/Security/ProviderConfiguration/Domain/OpenId/Exceptions/OpenIdConfigurationException.php]:
```suggestion
                . ' %s given',
```

File: src/Core/Security/Authentication/Infrastructure/Api/Login/OpenId/LoginController.php [hunk 2]
```diff
@@ -60,17 +62,26 @@ public function __invoke(
 
         $useCase($request, $presenter);
 
-        $response = $presenter->getPresentedData();
-        if ($response->getException() !== null) {
-            return View::createRedirect(
-                $this->getBaseUrl() . '/login?authenticationError=' . $response->getError()->getMessage()
-            );
+        switch (true) {
+            case is_a($presenter->getResponseStatus(), PasswordExpiredResponse::class)
```
Comment (kduret) [inline src/Core/Security/Authentication/Infrastructure/Api/Login/OpenId/LoginController.php:66]: Password expiration is managed for OpenId ?

File: src/Core/Security/Authentication/Domain/Exception/SSOAuthenticationException.php [hunk 3]
```diff
@@ -177,4 +177,14 @@ public static function unableToDecodeIdToken(): self
...
     }
+
+    /**
+     * Exception thrown when the request to authentication condition fail
+     *
+     * @return self
+     */
+    public static function requestForCustomAuthenticationConditionsEndpointFail(): self
+    {
+        return new self(_('Request for authentication conditions custom endpoint has failed'));
```
Comment (kduret) [inline src/Core/Security/Authentication/Domain/Exception/SSOAuthenticationException.php:188]: is it already translated ?

File: src/Core/Security/Authentication/Domain/Exception/AuthenticationConditionsException.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    /**
+     * Exceptions thrown when authentication are not found.
+     *
+     * @return self
+     */
+    public static function conditionsNotFound(): self
+    {
+        return new self(_("Authorized conditions not found in provider conditions"));
```
Comment (kduret) [inline src/Core/Security/Authentication/Domain/Exception/AuthenticationConditionsException.php:45]: is it already translated ?

File: src/Core/Security/Authentication/Application/UseCase/Login/LoginResponse.php [hunk 5]
```diff
@@ -50,4 +51,9 @@ public function getException(): ?Exception
     {
         return $this->exception;
     }
+
+    public function getMessage(): string
```
Comment (kduret) [inline src/Core/Security/Authentication/Application/UseCase/Login/LoginResponse.php]: phpdoc

File: src/Core/Security/Authentication/Application/UseCase/Login/Login.php [hunk 6]
```diff
@@ -124,9 +125,16 @@ public function __invoke(LoginRequest $loginRequest, PresenterInterface $present
...
             throw $e;
+        } catch (AuthenticationConditionsException $ex) {
+            $presenter->setResponseStatus(new ErrorAuthenticationConditionsResponse($ex->getMessage()));
+            return;
         }
 
+        $presenter->setResponseStatus(
+            new LoginResponse($this->getRedirectionUri($user, $loginRequest->refererQueryParameters))
+        );
         $presenter->present(new LoginResponse($this->getRedirectionUri($user, $loginRequest->refererQueryParameters)));
```
Comment (kduret) [inline src/Core/Security/Authentication/Application/UseCase/Login/Login.php:136]: note for later, we will need to improve the global presenter mechanism

File: src/Core/Security/Authentication/Domain/Exception/SSOAuthenticationException.php (hunk 3 above)
Comment (jeremyjaouen) [reply src/Core/Security/Authentication/Domain/Exception/SSOAuthenticationException.php:188]: indeed i'll add the translation

File: src/Core/Security/Authentication/Domain/Exception/AuthenticationConditionsException.php (hunk 4 above)
Comment (jeremyjaouen) [reply src/Core/Security/Authentication/Domain/Exception/AuthenticationConditionsException.php:45]: indeed i'll add the translation

File: src/Core/Security/Authentication/Infrastructure/Api/Login/OpenId/LoginController.php (hunk 2 above)
Comment (jeremyjaouen) [reply src/Core/Security/Authentication/Infrastructure/Api/Login/OpenId/LoginController.php:66]: No, but as the UseCase is Unique, the presenter in the useCase can present this response.

File: src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php [hunk 7]
```diff
@@ -537,15 +550,26 @@ private function sendRequestForIntrospectionTokenOrFail(): void
...
 
     /**
      * Send a request to get user information.
+     * @return array<string,mixed>
      * @throws SSOAuthenticationException
      */
-    private function sendRequestForUserInformationOrFail(): void
+    private function sendRequestForUserInformationEndpoint(): array
     {
         $this->info('Send Request for User Information...');
```
Comment (cg-tw) [inline src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php]:
```suggestion
        $this->info('Sending Request for User Information...');
```

File: src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php [hunk 8]
```diff
@@ -585,34 +609,32 @@ private function sendRequestForUserInformationOrFail(): void
...
     /**
      * Validate that Client IP is allowed to connect to external provider.
      *
      * @param string $clientIp
      * @throws SSOAuthenticationException
+     * @throws AuthenticationConditionsException
      */
     private function verifyThatClientIsAllowedToConnectOrFail(string $clientIp): void
     {
         $this->info('Check Client IP from blacklist/whitelist addresses');
```
Comment (cg-tw) [inline src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php]:
```suggestion
        $this->info('Check Client IP against blacklist/whitelist addresses');
```

File: src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php [hunk 9]
```diff
@@ -807,6 +831,21 @@ private function logAuthenticationInfo(string $message, array $content): void
         $this->debug('Authentication informations : ', $content);
```
Comment (cg-tw) [inline src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php]:
```suggestion
        $this->debug('Authentication information : ', $content);
```

File: src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php [hunk 10]
```diff
@@ -824,4 +863,197 @@ private function logExceptionInLoginLogFile(string $message, \Exception $e): voi
...
+    /**
+     * Send Request to get conditions for conditions custom endpoint.
+     *
+     * @param string $customEndpoint
+     * @return array<string,mixed>
+     * @throws SSOAuthenticationException
+     */
+    private function sendRequestForCustomAuthenticationConditionEndpoint(string $customEndpoint): array
+    {
+        $this->info('Send Request for authentication conditions...');
```
Comment (cg-tw) [inline src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php]:
```suggestion
        $this->info('Sending Request for authentication conditions...');
```

File: src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php [hunk 11]
```diff
@@ -824,4 +863,197 @@ private function logExceptionInLoginLogFile(string $message, \Exception $e): voi
...
+                foreach ($array as $k => $v) {
+                    if ($k !== $i++) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+        }
+        if (array_is_list($providerAuthenticationConditions) === false) {
+            $errorMessage = "Invalid Authentication conditions format, array of string expected";
```
Comment (cg-tw) [inline src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php]:
```suggestion
            $errorMessage = "Invalid Authentication conditions format, array of strings expected";
```

File: lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 12]
```diff
@@ -17042,4 +17042,10 @@ msgid "Invalid endpoint type specified"
 msgstr "Type d'endpoint spécifié invalide"
 
 msgid "Invalid endpoint URL specified"
-msgstr "URL d'endpoint spécifié invalide"
\ No newline at end of file
+msgstr "URL d'endpoint spécifié invalide"
+
+msgid "Request for authentication conditions custom endpoint has failed"
+msgtr "La demande de conditions d'authentification de l'endpoint personnalisé a échoué"
```
Comment (cg-tw) [inline lang/fr_FR.UTF-8/LC_MESSAGES/messages.po]:
```suggestion
msgtr "La demande d'endpoint personnalisé de conditions d'authentification a échoué"
```

**Review by kduret — APPROVED** [review]

**Review by adr-mo — APPROVED** [review]

**Pull request merged**